crates/subspace-block-relay/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+    /// Downloaded unit Id
+    download_unit_id: DownloadUnitId,
+
+    /// Downloaded entry
+    download_unit: Vec<u8>,
+
+    /// Total transactions (in bytes) that could not be resolved
+    /// locally, and had to be fetched from the server
+    local_miss: usize,
+
+    /// Download latency
+    latency: Duration,
+}
+
+/// The resolved protocol unit related info
+pub(crate) struct Resolved<ProtocolUnitId, ProtocolUnit> {
+    /// The protocol unit Id.
+    pub(crate) protocol_unit_id: ProtocolUnitId,
+
+    /// The protocol unit
+    pub(crate) protocol_unit: ProtocolUnit,
+
+    /// If it was resolved locally, or if it had to be
+    /// fetched from the server (local miss)
+    pub(crate) locally_resolved: bool,
+}
+
+/// The client side of the relay protocol
+#[async_trait]
+pub(crate) trait ProtocolClient<DownloadUnitId, ProtocolUnitId, ProtocolUnit>:
+    Send + Sync
+where
+    DownloadUnitId: Encode + Decode,
+    ProtocolUnitId: Encode + Decode,
+    ProtocolUnit: Encode + Decode,
+{
+    /// Builds the protocol portion of the initial request
+    fn build_initial_request(&self) -> Option<Vec<u8>>;
+
+    /// Resolves the initial response to produce the protocol units
+    async fn resolve(
+        &self,
+        response: Vec<u8>,
+        stub: Arc<dyn NetworkStub>,
+    ) -> Result<(DownloadUnitId, Vec<Resolved<ProtocolUnitId, ProtocolUnit>>), RelayError>;
+}
+
+/// The server side of the relay protocol
+pub(crate) trait ProtocolServer<DownloadUnitId>
+where
+    DownloadUnitId: Encode + Decode,
+{
+    /// Builds the protocol response to the initial request
+    fn build_initial_response(
+        &self,
+        download_unit_id: &DownloadUnitId,
+        initial_request: Option<Vec<u8>>,
+    ) -> Result<Vec<u8>, RelayError>;
+
+    /// Handles the additional client messages during the reconcile phase
+    fn on_request(&self, request: Vec<u8>) -> Result<Vec<u8>, RelayError>;
+}
+
+/// The relay user specific backend interface
+pub(crate) trait ProtocolBackend<DownloadUnitId, ProtocolUnitId, ProtocolUnit>
+where
+    DownloadUnitId: Encode + Decode,
+    ProtocolUnitId: Encode + Decode,
+    ProtocolUnit: Encode + Decode,
+{
+    /// Returns all the protocol units for the given download unit
+    fn download_unit_members(
+        &self,
+        id: &DownloadUnitId,
+    ) -> Result<Vec<(ProtocolUnitId, ProtocolUnit)>, RelayError>;
+
+    /// Returns the protocol unit for the given download/protocol unit
+    fn protocol_unit(
+        &self,
+        download_unit_id: &DownloadUnitId,
+        protocol_unit_id: &ProtocolUnitId,
+    ) -> Result<Option<ProtocolUnit>, RelayError>;
+}
+
+/// Network interface helper
+#[async_trait]
+pub(crate) trait NetworkStub: Send + Sync {
+    /// Performs the request response and returns the result
+    async fn request_response(
+        &self,
+        request: Vec<u8>,
+        is_protocol_message: bool,
+    ) -> Result<Result<Vec<u8>, RequestFailure>, Canceled>;
+}
```

Comment:
You mean the name? This can be changed if you have suggestions

Why this is needed: so that the use case specific types (like `BlockT` are not leaked here), `NetworkStubImp` needs `<Block: BlockT>`

Reply:
Yes, primarily the name. Stub typically means it is something for testing purposes and such, it is clearly not here judging by the usage.

Reply:
Changed it to NetworkInterface, stub is the gRPC terminology for the client side handle